tools/test

```diff
@@ -82,6 +84,7 @@ while (( $# )); do
             esac
             shift
             ;;
+        --all) files=all; platform=both; shift;;
```

Comment:
I think it someone does `tools/test --all --platform=android` or similar, the `--platform` will overwrite the `--all`, is that right?

Probably we should just fail if someone passes both `--all` and `--platform` or `--diff`?

Reply:
Makes sense. I also wondered if we should write the first line of the usage string differently, to reflect that:

```
usage: tools/test [--all-files | --diff COMMIT] [--platform <ios|android|both|sloppy>] [--all] [SUITES]
```

(I.e., from reading that it looks like it's perfectly fine to do `--all` and `--platform` or `--diff`.)

But I'm not sure how best to do that while preserving readability. Maybe it's fine as-is.

Reply:
Hmm, looking again, I'm not sure what's the best way to enforce these rules on what gets passed. (I see that we're already not enforcing that only one of `--diff` and `--all-files` gets passed.)

I'm not super familiar with Bash, yet. It seems plausible to me that the best way to start enforcing these things will be after we've started using `getopt` for something different; Greg suggested that [here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.60jest-expo.2Fios.60.20and.20.60jest-expo.2Fandroid.60/near/1169550). Do you think that's right?

Reply:
For `--all-files` and `--diff`, I think the behavior is actually fine: they're each setting the same option to different values, and the last one wins. That's a common CLI convention, and it can come in handy when you want to have some default command line (e.g. an alias in your interactive shell) and sometimes pass additional arguments, and sometimes override an option that was already in the default version of the command.

I think the way `--all` interacts with those two and `--platform` is also fine for the same reason. It means that, for example, you could have an alias that expands to `tools/test --all`, and effectively it'd be `tools/test` but with your custom choice of a different default -- you'd still be able to override it to get other behavior when you wanted.

(For the real thorough version of that, we'd also add a flag that means the default `files=branch` behavior as opposed to `--all-files` or `--diff`.)

If we did want behavior where some options are mutually exclusive, that's actually orthogonal to getopt; the way you'd implement it if we were using getopt is basically the same as how you'd implement it now.